### PR TITLE
ossl_quic_new(): Fix a leak found by error injection

### PR DESCRIPTION
**Pull request by t8m**
Fix for 
https://github.com/openssl/openssl/pull/21668#issuecomment-1787221129

### CODE REVIEW

File: ssl/quic/quic_impl.c [hunk 1]
```diff
@@ -384,6 +384,12 @@ SSL *ossl_quic_new(SSL_CTX *ctx)
         QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_CRYPTO_LIB, NULL);
         goto err;
     }
+#if defined(OPENSSL_THREADS)
+    if ((qc->mutex = ossl_crypto_mutex_new()) == NULL) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_CRYPTO_LIB, NULL);
+        goto err;
+    }
+#endif
```
Comment (mattcaswell) [inline ssl/quic/quic_impl.c:392]: Why is this necessary? Doesn't this get freed by `SSL_free(ssl_base)`?

Comment (t8m) [reply ssl/quic/quic_impl.c:392]: The mutex is needed to be there in case we call SSL_free(ssl_base) as it is being locked when it calls ossl_quic_free(). With the current ordering it can be not yet allocated when ossl_ssl_connection_new_int() below fails.

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. Thank you for the reviews.
